File: frontend/src/api/history.ts
import api from './client'
import type { DeviceHistoryItem, CellHistoryItem, CellStabilityItem } from '@/types'

export interface DeviceHistoryParams {
  ctn?: string
  from?: string
  to?: string
}

export interface CellHistoryParams {
  ecgi?: number
  band?: number
  from?: string
  to?: string
}

export const fetchDeviceHistory = (params: DeviceHistoryParams) =>
  api.get<DeviceHistoryItem[]>('/history/devices', { params }).then((r) => r.data)

export const fetchCellHistory = (params: CellHistoryParams) =>
  api.get<CellHistoryItem[]>('/history/cells', { params }).then((r) => r.data)

export const fetchCellStability = () =>
  api.get<CellStabilityItem[]>('/history/cells').then((r) => r.data)

export const resetHistory = () =>
  api.delete<{ deleted: number; rows: number; bytes: number }>('/history/reset').then((r) => r.data)

export const fetchHistoryStats = () =>
  api.get<{ rows: number; bytes: number }>('/history/stats').then((r) => r.data)

File: cell_traffic_optimizer/src/cell_traffic_optimizer/server/routers/history.py
from __future__ import annotations
from typing import Optional
from fastapi import APIRouter, Depends, Query
from ..deps import get_state
from ..state import AppState
from ..schemas import DeviceHistoryItemSchema, CellHistoryItemSchema

router = APIRouter(prefix="/api/history", tags=["history"])


@router.get("/devices", response_model=list[DeviceHistoryItemSchema])
def get_device_history(
    ctn: Optional[str] = Query(None),
    from_: Optional[str] = Query(None, alias="from"),
    to: Optional[str] = Query(None),
    state: AppState = Depends(get_state),
):
    rows = state.event_store.query_device_history(
        ctn=ctn or None,
        from_ts=from_ or None,
        to_ts=to or None,
    )
    result = []
    for r in rows:
        try:
            result.append(DeviceHistoryItemSchema(**r))
        except Exception:
            pass
    return result


@router.get("/cells", response_model=list[CellHistoryItemSchema])
def get_cell_history(
    ecgi: Optional[int] = Query(None),
    band: Optional[int] = Query(None),
    from_: Optional[str] = Query(None, alias="from"),
    to: Optional[str] = Query(None),
    state: AppState = Depends(get_state),
):
    rows = state.event_store.query_cell_history(
        ecgi=ecgi,
        band=band,
        from_ts=from_ or None,
        to_ts=to or None,
    )
    result = []
    for r in rows:
        try:
            result.append(CellHistoryItemSchema(
                id=r["id"],
                timestamp=r["timestamp"],
                eventType=r["eventType"],
                ecgi=r["groupingKey"]["ecgi"],
                band=r["groupingKey"]["band"],
                message=r["message"],
            ))
        except Exception:
            pass
    return result


@router.delete("/reset")
def reset_history(state: AppState = Depends(get_state)):
    deleted = state.event_store.reset()
    size = state.event_store.size()
    return {"deleted": deleted, "rows": size["rows"], "bytes": size["bytes"]}


@router.get("/stats")
def history_stats(state: AppState = Depends(get_state)):
    return state.event_store.size()
